src/main/java/com/rackspace/jenkins_nodepool/NodePoolRequestStateWatcher.java

```diff
@@ -0,0 +1,154 @@
+
+        log.fine("Creating watch for state:" + desiredState + " on path:" + zpath);
+        // The watch will be triggered by a successful operation that creates/delete the node or sets the data on the node.
+        final Stat stat = curatorFramework.checkExists().usingWatcher(this).forPath(zpath);
+        if (stat == null) {
+            log.warning("Created watch on non-existent path: " + zpath);
+        }
+    }
+
+    /**
+     * Our event callback method.  When something changes and the request is fulfilled, we'll adjust our latch and the
+     * blocking call to waitUntilDone() will return.
+     *
+     * @param event the watch event value provided by the callback
+     */
+    @Override
+    public void process(WatchedEvent event) {
+        log.fine("Watch event received:" + event.toString());
+
+        try {
+            // Re-register if not a NodeDeleted and not a None event type
+            if (event.getType() != Watcher.Event.EventType.NodeDeleted && event.getType() != Watcher.Event.EventType.None) {
+                // Continuous watching on znodes requires reset/re-register of watches after every event/trigger.
+                // Need to quickly re-register as the pending -> fulfilled state transition may happen fast - don't want
+                // to miss the event.
+                registerWatch(zpath);
+            }
+
+            // Only interested in the node data changed event
+            if (event.getType() == Watcher.Event.EventType.NodeDataChanged) {
+
+                // Let's fetch the data from ZK directly so we can read the state value
+                final RequestState state = getStateFromPath();
+                log.fine("Watch event received event type:" + event.getType() + ". State is:" + state);
+
+                // Are we in the desired state yet? No reason to continue if the request failed.
+                if (state == desiredState || state == RequestState.failed) {
+                    latch.countDown();
+                } else {
+                    log.fine("Watch event ignoring event type:" + event.getType() +
+                            ". Fetched state is: " + state + "/" + getStateFromPath());
+                }
+            } else {
+                log.fine("Watch event ignoring event type:" + event.getType());
+            }
+        } catch (Exception e) {
+            log.warning(e.getClass().getSimpleName() +
+                    " occurred while updating state from Zookeeper. Message: " + e.getLocalizedMessage() +
+                    ". Ignoring the event.");
+        }
+    }
+
+    /**
+     * Waits until event fires or until the specified timeout.
+     *
+     * @param timeout the timeout value
+     * @param unit    the unit of the timeout value, typically TimeUnit.SECONDS
+     * @throws InterruptedException if the current thread is interrupted while waiting
+     */
+    public void waitUntilDone(long timeout, TimeUnit unit) throws InterruptedException {
```

Comment:
Add function annotation to avoid the [findbugs failure](https://ci.jenkins.io/blue/organizations/jenkins/Plugins%2Fnodepool-agents-plugin/detail/PR-4/6/pipeline/9#step-31-log-152) `@SuppressFBWarnings("RV_RETURN_VALUE_IGNORED")`

Reply:
On a void return method signature? Seems unnecessary, but ok.

Reply:
The error is on the latch.await call within waitUntilDone, not on waitUntilDone itself.